3단계 - @MVC 구현(힌트)

willing1026 opened the pull request:
피드백 부탁드립니다

1. step2 피드백 반영
   - HandlerAdapter 인터페이스 적용
  ServletHandlerAdapter를 생성하고 HandlerMappings 1급 컬랙션을 인스턴스로 갖도록 했습니다. (HalderMappings 는 LegacyMvcHandlerMapping 과 AnnotationHandlerMapping을 갖도록 했습니다.)

      step2에서 주셨던 피드백을 제대로 이해하고 반영한건지 궁금합니다.
      > 프레임워크를 구현하는 관점에서 봤을 때 Servlet, Struts1, Struts2와 같은 다양한 프레임워크와의 통합을 하고 싶다.
이와 같이 새로운 기술에 대한 통합을 시도할 경우 if/else가 계속해서 추가될 수 있다.
이와 같이 if/else가 늘어날 것으로 예상되는 부분에 자바의 다형성을 활용해 interface를 추가하고 확장 가능한 구조로 개선해 보면 어떨까?


2. 요구사항 1에 대한 수정
   - ControllerScanner 추가.

감사합니다 😄

willing1026 committed:
# add & modify : HandlerMapping 인터페이스 추가,  RequestMapping 적용

## add
- HandlerMapping : RequestMapping, AnnotationHandlerMapping 공통기능 추상화 (initialize, getHandler)

## modify
- LegacyMvcHandlerMapping : HandlerMapping 상속, 공통기능 정리, ,mappingList정리(중복제거), 이름변경( from RequestMapping )
- DispatcherServlet : LegacyMvcHandlerMapping 변경에 따른 기능수정
- 그 외 테스트케이스 추가 및 수정

willing1026 committed:
# modify - LegacyMvcHandlerMapping.getHandler return type 수정 ( Object -> Controller )

willing1026 committed:
# modify : handlerMapping 인터페이스 상속

willing1026 committed:
# modify : handlerMapping List에서 handler를 찾은 뒤 Controller 수행하도록 수정

- DispatcherServlet : 초기화 구문 수정(LegacyMvcMapping, AnnotationHandlerMapping 인스턴스변수 제거), handler 타입에 따른 실행방식 결정로직 추가
- HandlerMapping : support method 추가 (requestURI 를 지원하는지 확인)
- LegayMvcHandlerMapping, AnnotationHandlerMapping : support 기능 구현

willing1026 committed:
# add & modify : dispatcherServlet에 있던 controller 수행 기능 HandlerAdapter로 이관

## add
- HandlerAdapter : handlerApdater interface
- ServletHandlerAdapter : servletHander 타입에 따른 처리 기능 구현
- HandlerMappings : HandlerMapping List 1급 컬랙션

## modify
- DispatcherServlet : handler 관련기능 제거. ServletHandlerAdapter로 기능 이관
- 그 외 테스트케이스 추가

willing1026 committed:
# add & modify : CotrollerScanner 추가, AnnotationHandlerMapping 기능이관

## add
- CotrollerScanner : controller scan, controller 인스턴스 생성 및 map에 저장

## modify
- AnnotationHandlerMapping : 컨트롤러 스캔기능 제거

willing1026 commented:
`HandlerAdapter`에 대한 피드백을 제대로 이해못했던거 같습니다.
legacyHandlerAdapter 와 AnnotationHandlerAdapter로 분리해보겠습니다!

javajigi commented on src/main/java/core/mvc/tobe/ServletHandlerAdapter.java:
```diff
@@ -0,0 +1,62 @@
+public class ServletHandlerAdapter implements HandlerAdapter {
+    private static final String DEFAULT_REDIRECT_PREFIX = "redirect:";
+
+    private HandlerMappings handlerMappings;
```
HandlerMapping과 HandlerAdapter가 의존관계를 가지지 않도록 구현하는 것이 어떨까?

willing1026 replied:
의존관계 맺지 않도록 분리했습니다!

javajigi reviewed (changes requested):
어제 강의에서 보여줬던 코드도 있는데요. HandlerAdapter를 분리해서 진행해 보면 좋을 것 같아요.
또한 HandlerMapping과 HandlerAdapter의 의존 관계는 분리해 보세요.

willing1026 committed:
# add, modify, delete : handlerApdater 개선 ( handlerMapping과의 의존관계 분리, legacy와 annotation adapter 추가 )

## delete
- ServletHandlerAdapter : LegacyHandleAdapter, AnnotationHandlerAdapter로 분리

## modify
- DispatcherServlet :인스턴스 변수 변경 ( ServletAdapters -> handlerApdaters )
- HandlerApdater : support method 인자를 handler를 받도록 수정
- HandlerMappings : support method제거, getHandler에서 예외처리하도록 수정, 그 외 인스턴스변수 이름 수정
- LegacyHandlerMapping : LegacyMvcHandlerMapping 에서 이름 변경 (불필요한 단어 제거)

## add
- LegacyHandlerAdapter, AnnotationHandlerApdater : ServletHanderAdapter 기능 분리한 대상들
- HandlerAdapters : handlerAdapter list의 1급 컬랙션

- 그 외 테스트케이스 추가 및 이름수정

willing1026 commented:
- HandlerAdapter와 HandlerMapping간의 의존관계 제거
- ServletHandlerAdapter를 LegacyHandlerAdpater 와 AnnotationHandlerAdapter로 분리

확인부탁드립니다 😄

javajigi reviewed (approved):
피드백 반영 잘 했네요. 💯 
Advanced로 빠져있는 "@MVC - Controller 메소드 인자 매핑" 미션 진행할 의사가 있으면 계속 진행하면 되고요.
지금 단계에서 미션 마무리하려면 DM으로 이야기해 주세요.
